download_py_package.py

```diff
@@ -35,6 +38,30 @@
 
 def download_py_package():
     """Download package which needs to be analyzed by future steps."""
+    page = request.urlopen(os.path.join(Configuration.PACKAGE_INDEX, Configuration.PACKAGE_NAME))
+    version_exists = False
+
+    for link in BeautifulSoup(
+        page, "html.parser", from_encoding=page.info().get_param("charset"), parse_only=SoupStrainer("a")
+    ):
+        if link.string.endswith(f"-{Configuration.PACKAGE_VERSION}.zip") or link.string.endswith(
+            f"-{Configuration.PACKAGE_VERSION}.tar.gz"
+        ):
+            break
+        elif f"-{Configuration.PACKAGE_VERSION}-" in link.string:
+            version_exists = True
+    else:
+        if version_exists:
+            raise SourceDistroNotFound(
+                f"No source distro found for {Configuration.PACKAGE_NAME}==={Configuration.PACKAGE_VERSION} on"
+                f" {Configuration.PACKAGE_INDEX}"
+            )
+        else:
+            raise VersionDoesNotExist(
+                f"Version not found for {Configuration.PACKAGE_NAME}==={Configuration.PACKAGE_VERSION}"
+                f" on {Configuration.PACKAGE_INDEX}"
+            )
+
```

Comment:
I think we have to save this information in the database, as `si_unresolved` would always find that a package version index is not SI analyzed and schedule SI workflow, we need to add a flag in the database, `is_downloadable`. It would be probably maybe good to modify SI workflow:

1. download package and do not fail if package cannot be download, but kindly end storing a flag `is_downloadable`

1a. check if `is_downloadable` is True then do nothing move to 2, otherwise create an output document for SI workflow where we state that the information cannot be downloaded and move to 2a.

1b is basically `graph-sync` but with logic in storages modified that check a flag in the SI workflow output document and if `is_downloadable` False it creates an entry in SI table with that flag set accordingly, so that we can modify queries filtering for that and avoiding rescheduling packages

2. si_bandit
3. si_cloc
4. si_aggreagate
5. graph-sync

To summarize if this solution looks good for you:

- [ ] workflow-helpers task do not fail in that case but kindly exit with flag that can trigger optionally other tasks
- [ ] modify SI workflow with new option and structure (check_is_downloadable task + graph-sync)
- [ ] modify sync_si in storages to check  for `is_downloadable` in the SI document
- [ ] modify queries to filter for that flag `is_downloadable`

WDYT @fridex @KPostOffice @goern

Reply:
The alternative I have in mind is to create a Kafka message and let investigator assign that flag in the database, similar to what it does for missing package from messages coming from package_update- job.

In this case tasks above are slightly modified:

- no modification in SI sync
- 1a stores a message
- no extra graph-sync, but instead 1b take the message and send-message task to Kafka

new tasks
- create  `PackageNotDownloadbaleMessage`
- add logic to investigator to handle `PackageNotDownloadbaleMessage`

Reply:
I think storing this as a flag by sending a message would be good.  For the workflow I think we can leave it as is and instead add some logic in common to check the flag before scheduling the workflow. I don't think making an entire workflow step to do a single DB query is a good idea.

Reply:
> I think storing this as a flag by sending a message would be good. 

Sounds good

For the workflow I think we can leave it as is and instead add some logic in common to check the flag before scheduling the workflow. 

common cannot use storages because storages uses common we have to do that on the workflow level

I don't think making an entire workflow step to do a single DB query is a good idea.

 I think as we do for solver workflow to retrigger adviser. The flexibility of the workflows make it quite is to do that, it's a task in a workflow. Graph-sync uses only one method of DB as well :) 

WDYT?

Reply:
What are we gaining by adding steps though.  If we still have to start the workflow and check if the package has gone missing what difference does it make whether we throw an error in DownloadPackage or a previous step.

Reply:
> What are we gaining by adding steps though. If we still have to start the workflow and check if the package is downloadable what difference does it make whether we throw an error in DownloadPackage or a previous step.

it won't be scheduled, because the query used to find SI unanalyzed will trigger only the one with the flag set to False. But SI workflow at the first round need to fail so that a flag can be created, otherwise we cannot know that in advance. Therefore SI need to have a logic that send a message for investigator to create that flag. So next round to schedule SI workflow that won't be rescheduled.

Reply:
So let's proceed as we said, here if the task is failing due to known issues, we store a parameter called `is_downloadable` set to 0, by default it will set to 1 and stored instead, so next tasks can run if that flag is set to 1 otherwise nothing.

Reply:
https://github.com/thoth-station/thoth-application/issues/507